#!/usr/bin/env node

/**
 * Module dependencies.
 */

const createExpressApp = require('../src/app');
const debug = require('debug')('supplymgmt:server');
const http = require('http');

/**
 * Get port from environment and store in Express.
 */
const app = createExpressApp();
const port = normalizePort(process.env.PORT || '4000');
app.set('port', port);

/**
 * Create HTTP server.
 */

const server = http.createServer(app);

/**
 * Listen on provided port, on all network interfaces.
 */

server.listen(port);
server.on('error', onError);
server.on('listening', onListening);

/**
 * Normalize a port into a number, string, or false.
 */

function normalizePort(val) {
  const port = parseInt(val, 10);

  if (isNaN(port)) {
    // named pipe
    return val;
  }

  if (port >= 0) {
    // port number
    return port;
  }

  return false;
}

/**
 * Event listener for HTTP server "error" event.
 */

function onError(error) {
  if (error.syscall !== 'listen') {
    throw error;
  }

  const bind = typeof port === 'string'
    ? 'Pipe ' + port
    : 'Port ' + port;

  // handle specific listen errors with friendly messages
  switch (error.code) {
    case 'EACCES':
      console.error(bind + ' requires elevated privileges');
      process.exit(1);
      break;
    case 'EADDRINUSE':
      console.error(bind + ' is already in use');
      process.exit(1);
      break;
    default:
      throw error;
  }
}

/**
 * Event listener for HTTP server "listening" event.
 */

function onListening() {
  const addr = server.address();
  const bind = typeof addr === 'string'
    ? 'pipe ' + addr
    : 'port ' + addr.port;
  debug('Listening on ' + bind);
  console.log('Listening on ' + bind);
}

process.on('uncaughtException', (err) => {
    console.error('There was an uncaught exception:', err.message);
    console.error(err.stack);
  
    // Attempt a graceful shutdown
    server.close(() => {
      console.log('Server closed due to uncaught exception.');
      process.exit(1); // Exit the process with an error code
    });
  
    // If the server fails to close after a certain timeout, force exit
    setTimeout(() => {
      console.error('Forcing shutdown due to uncaught exception.');
      process.exit(1);
    }, 10000).unref(); // unref to allow the process to exit if there are no other pending operations
  });
